tikv-in-memory-engine.md

```diff
@@ -0,0 +1,131 @@
+
+## Working principles
+
+TiKV MVCC in-memory engine caches the latest written MVCC versions in memory and implements an MVCC GC mechanism independent of TiDB, allowing it to quickly perform GC on MVCC versions in memory, reducing the number of versions scanned during queries, and achieving the effect of reducing request latency and CPU overhead.
+
+The following diagram illustrates how TiKV organizes MVCC versions.
+
+<div style="text-align: center;"><img src="./media/tikv-ime-data-organization.png" alt="IME caches recent versions to reduce CPU overhead" width="400" /></div>
+
+The diagram shows two rows of records, each with 9 MVCC versions. The behavior is compared between the cases with and without IME enabled:
+
+- On the left, without IME enabled, the table records are stored in RocksDB in ascending order by primary key, with the same row's MVCC versions adjacent to each other.
+- On the right, with IME enabled, the data in RocksDB is consistent with the left side, and IME caches the latest 2 MVCC versions of the 2 rows of records.
+- When TiKV processes a scan request with a range of `[k1, k2]` and a start timestamp of `8`, the left side without IME enabled needs to process 11 MVCC versions, while the right side with IME enabled only needs to process 4 MVCC versions, reducing request latency and CPU consumption.
+- When TiKV processes a scan request with a range of `[k1, k2]` and a start timestamp of `7`, because the right side lacks the historical versions that need to be read, the IME cache becomes invalid, and it falls back to reading data from RocksDB.
+
+## Usage
+
+Enabling IME requires adjusting the TiKV configuration and restarting. The following example explains the configuration items:
+
+```toml
+[in-memory-engine]
+# This parameter is the switch for the in-memory engine feature, which is disabled by default. You can set it to true to enable it.
+enable = false
+#
+# This parameter controls the memory capacity that in-memory engine can use. The default value is 10% of the system memory, and the maximum value is 5 GiB.
+# You can manually configure it to use more memory.
+# Note: When in-memory engine is enabled, block-cache.capacity will be reduced by 10%.
+#capacity = "5GiB"
+#
+# This parameter controls the time interval for in-memory engine to GC the cached MVCC versions.
+# The default value is 3 minutes, representing that GC is performed every 3 minutes on the cached MVCC versions.
+# Decreasing the value of this parameter can speed up the GC frequency, reduce MVCC versions, but will increase GC CPU consumption and increase the probability of cache miss.
+#gc-run-interval = "3m"
+#
+# This parameter controls the threshold for in-memory engine to select and load Regions based on MVCC read amplification.
+# The default value is 10, indicating that when the number of MVCC versions processed for a row of records in a Region exceeds 10, it might be loaded into the in-memory engine.
+#mvcc-amplification-threshold = 10
```

Comment:
```suggestion
enable = false

# This parameter controls the memory capacity that in-memory engine can use. The default value is 10% of the system memory, and the maximum value is 5 GiB.
# You can manually configure it to use more memory.
# Note: When in-memory engine is enabled, block-cache.capacity will be reduced by 10%.
capacity = "5GiB"

# This parameter controls the time interval for in-memory engine to GC the cached MVCC versions.
# The default value is 3 minutes, representing that GC is performed every 3 minutes on the cached MVCC versions.
# Decreasing the value of this parameter can speed up the GC frequency, reduce MVCC versions, but will increase GC CPU consumption and increase the probability of cache miss.
gc-run-interval = "3m"

# This parameter controls the threshold for in-memory engine to select and load Regions based on MVCC read amplification.
# The default value is 10, indicating that when the number of MVCC versions processed for a row of records in a Region exceeds 10, it might be loaded into the in-memory engine.
mvcc-amplification-threshold = 10
```